### PR TITLE
Fixes #32321 - allow dualstack provisioning

### PR DESCRIPTION
**Pull request by lzap**
A customer reports that dual-stack provisioning does work. IPv4 is used for provisioning. We do not have capacity to test those workflows at the moment, but we can actually allow provisioning in these cases via a host parameter to determine which stack to configure for provisioning.

@ekohl

### CODE REVIEW

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb [hunk 1]
```diff
@@ -32,10 +32,10 @@ snippet: true
   options.push("kssendmac", "ks.sendmac", "inst.ks.sendmac")
 
   # networking credentials
-  raise("Dual-stack provisioning not supported") if subnet4 && subnet6
-  if subnet4 && subnet4.dhcp_boot_mode?
+  raise("Both IPv4 and IPv6 subnets are set on a provisioning interface, create 'dualstack' parameter with either 'IPv4' or 'IPv6' to identify which to use.") if subnet4 && subnet6 && !host_param("dualstack").present?
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb]: Minor language suggestions:
```suggestion
  raise("Both IPv4 and IPv6 subnets are set on a provisioning interface, create a 'dualstack' parameter with either 'IPv4' or 'IPv6' to identify which protocol to use.") if subnet4 && subnet6 && !host_param("dualstack").present?
```

However, I still think this is a confusing name. When I read dual stack, I think both IPv4 and IPv6. Thoughts on using `kickstart-protocol=force-ipv6` instead of `dualstack=IPv6`?

**Review by ekohl — COMMENTED** [review]
I still think we should support real dual stack with both IPv4 and IPv6.

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb (hunk 1 above)
Comment (lzap) [reply app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb]: Amended the change, thanks.

**Comment by lzap** [comment]
> I still think we should support real dual stack with both IPv4 and IPv6.

What's real dualstack?

**Comment by ekohl** [comment]
> What's real dualstack?

Dual stack means having both stacks active at the same time. At least on the host. Applications can then decide stack to use.

**Comment by lzap** [comment]
IIRC this works already today. Both stacks should be configured.

**Comment by ekohl** [comment]
That's not how I read it. Just taking the if statements:
```ruby
if subnet4 && subnet4.dhcp_boot_mode? && host_param("dualstack") != "IPv6"
elsif subnet4 && !subnet4.dhcp_boot_mode? && host_param("dualstack") != "IPv6"
elsif subnet6.dhcp_boot_mode? && rhel_compatible && major >= 7
elsif subnet6 && !subnet6.dhcp_boot_mode?
end
```

That means it is not configuring both stacks but only one. You just allow choosing which one.

What I would expect a dual stack to look like is:
```ruby
if subnet4
  if subnet4.dhcp_boot_mode?
  else
  end
end

if subnet6
  if subnet6.dhcp_boot_mode? && rhel_compatible && major >= 7
  else
  end
end
```

Now I understand that in practice it's probably not so easy and it may be too difficult for now, but I'd at least agree on definitions.

**Comment by lzap** [comment]
I think we are talking apples and oranges.

This file I am changing configures Anaconda installer (or dracut to be precise). That configuration will disappear once OS is rebooted, for that we have a set of templates which deploy ifcfg configurations for NetworkManager. And IIRC we do configure both IPv4 and IPv6 in there.

IIUC dracut will only support either IPv4 or IPv6.

**Comment by ekohl** [comment]
I disagree and it's crucial for terminology to be used correctly. Other examples where it may be important are external services. Let's say you have one resource (like a yum rep) that's only available over IPv4 but another (like Foreman itself) is only available over IPv6. Then it's important that Anaconda is also configured dual stack because otherwise kickstart may fail. If you start to call this dual stack, an admin will be very frustrated after debugging it for a while.

So in short: I care that we use good naming to set the right expectations for users. Dual stack is having both stacks enabled at the same time.

**Comment by lzap** [comment]
> Dual stack is having both stacks enabled at the same time.

I agree, and this patch enables users to actually provision systems with dualstack configuration. Before, we errored out requiring user to use one or another. It is the right term, with one "but": installation is only performed via a single stack (which is the one user needs to pass as a host parameter).

**Comment by lzap** [comment]
If this is all about the parameter name "dualstack" I have no issue renaming it. Any ideas?

**Comment by ekohl** [comment]
That's what I suggested in https://github.com/theforeman/foreman/pull/8447#discussion_r613994749:
> However, I still think this is a confusing name. When I read dual stack, I think both IPv4 and IPv6. Thoughts on using `kickstart-protocol=force-ipv6` instead of `dualstack=IPv6`?

**Comment by lzap** [comment]
Oh it was under the suggestion and I missed it. I do not like having "kickstart" in it since we can implement the same for other OSes at some point and we would need to introduce a new name. So changing to `install-force=IPv6`.

**Review by ekohl — CHANGES_REQUESTED** [review]
I dislike `install-force` since it is very unspecific. You also missed 2 of the old usages.

**Comment by lzap** [comment]
I am going with "install-protocol".

When I was working on this, I found that the new snapshot tests were not working correctly, all hosts were actually IPv4 not IPv6. I am making a change (in a separate commit) that moves all stubs into factories, everything is cleaner now and subnets are finally correct.

**Comment by lzap** [comment]
Rebased, I made huge improvements. Please do take a look @ekohl

File: app/services/foreman/template_snapshot_service.rb [hunk 2]
```diff
@@ -37,85 +37,39 @@ def templates
...
-      host = FactoryBot.build(:host_for_snapshots_ipv4_dhcp_el7,
-        name: 'snapshot-ipv4-6-dhcp-el7',
-        subnet: FactoryBot.build(:subnet_ipv4_dhcp_for_snapshots),
-        subnet6: FactoryBot.build(:subnet_ipv6_dhcp_for_snapshots),
-        interfaces: [ipv46_interface])
-      define_host_params(host)
+      FactoryBot.build(:snapshot_host,
+        :with_snapshot_dhcp_dualstack,
+        :with_snapshot_os_el7,
+        :with_snapshot_puppet)
```
Comment (lzap) [inline app/services/foreman/template_snapshot_service.rb:80]: Big changes here, I think it was you who recommended this in the first place.

File: app/views/unattended/provisioning_templates/snippet/header.erb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+# Primary IPv4: <%= @host.primary_interface.ip || "unset" %>
+# Provision IPv6: <%= @host.provision_interface.ip6 || "unset" %>
+# Primary IPv6: <%= @host.primary_interface.ip6 || "unset" %>
+<%
+if plugin_present?('katello')
+-%>
+# Lifecycle environment: <%= @host.lifecycle_environment %>
+# Content View: <%= @host.content_view %>
+# Content Source: <%= @host.content_source %>
+<% end -%>
```
Comment (lzap) [inline app/views/unattended/provisioning_templates/snippet/header.erb:25]: Lack if visibility in snapshot template made us actually merging something that was not working. Previously, all hosts were using IPv4 and not IPv6 at all. This header allows us to directly see what snapshots (and templates) were rendered with. See below.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub2/Kickstart_default_PXEGrub2.host4and6dhcp.snap.txt [hunk 4]
```diff
@@ -1,10 +1,22 @@
...
+# Location: Location 1
+# Interfaces: 2
+# IPv4 subnet: snapshot-ipv4-dhcp
+# IPv6 subnet: snapshot-ipv6-dhcp
+# Provision interface: 00-f0-54-1a-7e-e4
+# Primary interface: 00-f0-54-1a-7e-e4
+# Provision IPv4: 192.168.42.42
+# Primary IPv4: 192.168.42.42
+# Provision IPv6: 2001:db8:42::42
+# Primary IPv6: 2001:db8:42::42
```
Comment (lzap) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/PXEGrub2/Kickstart_default_PXEGrub2.host4and6dhcp.snap.txt:13]: I think this is extremely useful not just for snapshots but overall.

**Comment by lzap** [comment]
@ekohl any other comments? I would love to get this merged and do some more snapshot tests work

**Comment by jcpunk** [comment]
I'll confess interest in this patch

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb [hunk 5]
```diff
@@ -36,11 +36,10 @@ snippet: true
   end
 
   # networking credentials
-  raise("Dual-stack provisioning not supported") if subnet4 && subnet6
-  options.push("dualstack!") if subnet4 && subnet6
-  if subnet4 && subnet4.dhcp_boot_mode?
+  raise("Both IPv4 and IPv6 subnets are set on a provisioning interface, create a 'install-protocol' parameter with either 'force-ipv4' or 'force-ipv6' to identify which protocol to use.") if subnet4 && subnet6 && !host_param("install-protocol").present?
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb]:
```suggestion
  raise("Both IPv4 and IPv6 subnets are set on a provisioning interface, create an 'install-protocol' parameter with either 'force-ipv4' or 'force-ipv6' to identify which protocol to use.") if subnet4 && subnet6 && !host_param("install-protocol").present?
```

File: app/views/unattended/provisioning_templates/snippet/header.erb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<%#
+kind: snippet
+name: header
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/snippet/header.erb:3]: I think we can find a better name for this.
```suggestion
name: print_context_header
```

File: test/factories/subnet.rb [hunk 7]
```diff
@@ -138,7 +138,7 @@
       gateway { '2001:db8:42::1' }
       dns_primary { '2001:db8:42::8' }
       dns_secondary { '2001:db8:42::4' }
-      mtu { '1442' }
+      mtu { '1142' }
```
Comment (ekohl) [inline test/factories/subnet.rb]: [RFC 2460](https://datatracker.ietf.org/doc/html/rfc2460#section-5) states:
> IPv6 requires that every link in the internet have an MTU of 1280 octets or greater.

**Review by ekohl — COMMENTED** [review]
Looks like this fell off my radar. Some small things inline, but overall :+1:

**Review by ekohl — CHANGES_REQUESTED** [review]
Could you take a look at my comments and resolve conflicts?

**Comment by ezr-ondrej** [comment]
What is the status here? If there will be no activity for two more weeks, we will close, but you're always welcome to reopen.

File: test/factories/subnet.rb (hunk 7 above)
Comment (lzap) [reply test/factories/subnet.rb]: I haven't introduced this but sure.

File: app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb (hunk 5 above)
Comment (lzap) [reply app/views/unattended/provisioning_templates/snippet/kickstart_kernel_options.erb]: Fixed.

File: app/views/unattended/provisioning_templates/snippet/header.erb (hunk 6 above)
Comment (lzap) [reply app/views/unattended/provisioning_templates/snippet/header.erb:3]: I am not changing this, I had terrible conflicts and need to wrap up today.

**Comment by lzap** [comment]
Rebased, resolved ton of conflicts please re-review carefully.
